File: zone-server/src/main/kotlin/net/bestia/zone/dialog/conversation/ConversationSMSG.kt
package net.bestia.zone.dialog.conversation

import net.bestia.bnet.proto.ConversationSmsgProto
import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.zone.dialog.toBnet
import net.bestia.zone.message.SMSG
import net.bestia.zone.util.EntityId

/**
 * Somebody is talking to you, and these are the things you may say back.
 *
 * A plain [SMSG] and deliberately **not** an `EntitySMSG`: the client routes those through its entity
 * handling and would try to build an entity for the id. The speaker is metadata here, the same choice
 * `DialogSMSG.sourceEntityId` makes.
 */
data class ConversationSMSG(
  val speakerEntityId: EntityId,
  val speakerName: String,
  val node: ConversationNode,
) : SMSG {

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val conversation = ConversationSmsgProto.ConversationSMSG.newBuilder()
      .setSpeakerEntityId(speakerEntityId)
      .setSpeakerName(speakerName)
      .setSpeech(node.speech.toBnet())

    node.options.forEach { conversation.addOptions(it.toBnet()) }

    return EnvelopeProto.Envelope.newBuilder()
      .setConversation(conversation.build())
      .build()
  }

  private fun Line.toBnet(): ConversationSmsgProto.Line {
    val line = ConversationSmsgProto.Line.newBuilder().setKey(key)
    args.forEach { (name, arg) -> line.putArgs(name, arg.toBnet()) }

    return line.build()
  }

  private fun ConversationOption.toBnet(): ConversationSmsgProto.ConversationOption {
    return ConversationSmsgProto.ConversationOption.newBuilder()
      .setTopicId(topicId)
      .setLine(line.toBnet())
      .setKind(kind.toBnet())
      .build()
  }

  // No `else`: a new OptionKind must fail to compile here rather than serialize as TALK and leave a
  // player clicking an option that silently does nothing.
  private fun OptionKind.toBnet(): ConversationSmsgProto.OptionKind = when (this) {
    OptionKind.TALK -> ConversationSmsgProto.OptionKind.TALK
    OptionKind.ACTION -> ConversationSmsgProto.OptionKind.ACTION
    OptionKind.BACK -> ConversationSmsgProto.OptionKind.BACK
    OptionKind.END -> ConversationSmsgProto.OptionKind.END
  }
}

File: bestia-client/src/Bnet/Message/System/ConversationSMSG.cs
using Godot;

namespace BestiaBehemothClient.Bnet.Message.System
{
  /// <summary>
  /// Somebody is talking to us, and these are the things we may say back.
  ///
  /// Account-scoped like <see cref="DialogSMSG"/> and deliberately an <see cref="ISMSG"/> rather
  /// than an EntitySMSG: the speaker is metadata, and routing this through entity handling would
  /// have the client try to build an entity for the id.
  ///
  /// <see cref="SpeakerName"/> is carried outright because entities have no display name on the
  /// wire yet. When they do, this field goes and the name comes off the entity.
  /// </summary>
  [GlobalClass]
  public partial class ConversationSMSG : ISMSG
  {
    [Export]
    public ulong SpeakerEntityId { get; set; } = 0;

    [Export]
    public string SpeakerName { get; set; } = string.Empty;

    [Export]
    public ConversationLine Speech { get; set; } = new();

    [Export]
    public Godot.Collections.Array<ConversationOption> Options { get; set; } = new();

    public static ConversationSMSG FromProto(global::Bnet.ConversationSMSG proto)
    {
      var options = new Godot.Collections.Array<ConversationOption>();
      foreach (var option in proto.Options)
      {
        options.Add(ConversationOption.FromProto(option));
      }

      return new ConversationSMSG
      {
        SpeakerEntityId = proto.SpeakerEntityId,
        SpeakerName = proto.SpeakerName ?? string.Empty,
        Speech = ConversationLine.FromProto(proto.Speech),
        Options = options
      };
    }
  }
}
